background/rust.tex

```diff
@@ -35,33 +86,56 @@ \subsubsection{Sum types - Enum}
   \label{lst:rust:list}
 \end{listing}
 
-Each list item is either a value and a pointer to the next item, as the \textbf{Cons}, or a termination value as \textbf{Nil}.
+\subsubsection{Structs}
+\label{ssub:structs}
 
-\subsubsection{Case Study - Option}
+\texttt{structs} in Rust are similar to that of C, and is a way of creating more advanced data types than \texttt{primitives} or \texttt{enums}.
+With a \texttt{struct} it is possible to combine multiple variables into a single type, each of them identified with its own name.
+An example of a \texttt{struct} is shown in \autoref{lst:book_struct}, everyone that knows C will see the similarity of the \texttt{struct} definition, but there are also a couple of extra things to notice from the example.
+Everything in Rust is \emph{private} by default, the \texttt{pub} keywords on type definitions and member fields make them \emph{publicly accessible}.
+The \texttt{impl} keyword lets us implement member functions for the \texttt{Book}, similar to class-declarations in C++ and other object-oriented languages.
 
 \begin{listing}[H]
   \begin{minted}{rust}
-    pub enum Option<T> {
-      Some(T),
-      None,
+    pub struct Book {
+      pub name: String,
+      pub pages: u32,
+    }
+
+    impl Book {
+      pub fn info(&self) {
+        println!("{} has {} pages.", self.name, self.pages);
+      }
     }
   \end{minted}
-  \caption{Definition of Option}
-  \label{lst:rust:option}
+  \caption{Struct definition and implementation}
+  \label{lst:book_struct}
 \end{listing}
 
-\subsubsection{Pattern Matching and Destruction}
+\subsubsection{Pattern Matching and Destructuring}
+\label{sub:pattern_matching}
+
+Pattern matching is a powerful language construct of Rust that resembles the \texttt{switch} statement from C.
+It is possible to match against any value or variable in the language, and branch to different blocks of code based on the outcome of the \texttt{match}.
+Another important feature of the \texttt{match} statement is that it needs to be \emph{exhaustive}, this means that all possible outcomes of a \texttt{match} need to be considered, otherwise Rust will make it a compile-time error.
+This is contrary to C, where only primitive values can be used in a \texttt{switch} statement, and there is no requirement to cover all switch cases, e.g. with the \texttt{default} keyword.
```

Comment:
Reword to this: `... e.g. the \texttt{default} keyword is not required.` ?

But this kind of reduces the Rust mechanism as it is a compile time check for validating that you have covered all possible values. Perhaps we should point to an example of refactoring an library by adding a variant to an `enum` causing the compiler to ensure that you add cases to all your matches ?